### PR TITLE
Supports censor function

### PR DESCRIPTION
**Pull request by feugy**
As discussed on slack, here is a first attempt to support censor functions.

The strategy I've used was to check the `censor` type in the beginning, store the information in `state`, and use it later to reduce the overhead as much as possible when serializing values.
I spot only 3 places to change:
- nominal case: redactor's `redactTmpl()` 
- trailing wildcard: modifiers' `groupRedact()`
- nested wildcard: modifiers' `specialSet()` used in `nestedRedact()`

I've added several tests everything looks fine.
But I'm not clear on `precensored` and the equality tests [here](https://github.com/davidmarkclements/fast-redact/blob/master/lib/redactor.js#L55) and [here](https://github.com/davidmarkclements/fast-redact/blob/master/lib/redactor.js#L46). How do they relate to this feature?

Finally, please note this is a breaking change, since the validator will not complain any more about `censor` being a function.

### CODE REVIEW

**Comment by mcollina** [comment]
Can you please ignore `yarn.lock`  and remove it from this PR? Thanks!

**Comment by davidmarkclements** [comment]
Overall I think this is nicely done 

I don't think we're saving anything by caching typeof censor and then passing it through all the state as censorFct


1. Could you just do a typeof check instead of using censorFct - and use benchmarks to verify no difference
1. Could you add a benchmark for censor as a function to show the cost of using a censor function 
1. Could you please run the benchmarks vs master and post results

**Comment by feugy** [comment]
Changes applied. Indeed, caching the test result doesn't bring significant improvement.

But to be honest, I'm not comfortable with the benchmark: the variation is quite big between 2 runs on the same branch (at least, on my machine: node 8.11.1, Linux Mint, Intel Core i7, 16Go RAM).
I could choose the results and make them saying the opposite :worried: 

Here are the (most honest) results:


Test | 1 – Censor Fct with cache | 2 – Censor Fct | 1-2 Δ | 3 – No support | 2-3 Δ
-- | -- | -- | -- | -- | --
Noir V2 | 88,73 | 89,98 | -1,26% | 88,308 | 1,68%
Fast Redact | 9,06 | 8,48 | 0,58% | 8,43 | 0,05%
Fast Redact Restore | 20,47 | 18,66 | 1,81% | 21,789 | -3,13%
Noir V2 Wild | 119,99 | 115,74 | 4,25% | 114,409 | 1,33%
Fast Redact Wild | 29,69 | 31,77 | -2,08% | 29,769 | 2,00%
Fast Redact Wild Restore | 34,67 | 35,08 | -0,41% | 38,238 | -3,15%
Fast Redact Intermediate Wild | 140,82 | 140,62 | 0,19% | 151,134 | -10,51%
Fast Redact Intermediate Wild Restore | 138,61 | 135,05 | 3,56% | 136,816 | -1,76%
JSON Stringify | 203,99 | 207,71 | -3,71% | 202,055 | 5,65%
Noir V2 Serialize | 280,60 | 286,06 | -5,46% | 266,796 | 19,26%
Fast Redact Serialize | 205,79 | 222,69 | -16,90% | 210,955 | 11,74%
Noir V2 Wild Serialize | 280,27 | 315,19 | -34,91% | 274,718 | 40,47%
Fast Redact Wild Serialize | 256,33 | 268,32 | -11,99% | 254,623 | 13,70%
Fast Redact Intermediate Wild Serialize | 325,63 | 355,86 | -30,22% | 330,755 | 25,10%
Fast Redact Intermediate Wild Match Wild Outcome Serialize | 476,18 | 495,73 | -19,56% | 468,75 | 26,98%
Fast Redact Static Match Wild Outcome Serialize | 263,65 | 225,00 | 38,65% | 244,574 | -19,57%
Noir V2 Censor Function | 84,53 | 81,29 | 3,25% | N/A | N/A
Fast Redact Censor Function | 56,76 | 52,27 | 4,49% | N/A | N/A
Noir V2 | 73,29 | 72,04 | 1,25% | 80,884 | -8,85%
Fast Redact | 2,03 | 2,02 | 0,02% | 1,949 | 0,07%
Fast Redact Restore | 12,63 | 13,67 | -1,04% | 10,72 | 2,95%
Noir V2 Wild | 69,83 | 87,60 | -17,77% | 71,49 | 16,11%
Fast Redact Wild | 22,50 | 24,26 | -1,76% | 20,798 | 3,47%
Fast Redact Wild Restore | 26,68 | 31,89 | -5,21% | 27,004 | 4,89%
Fast Redact Intermediate Wild | 155,84 | 197,66 | -41,82% | 161,807 | 35,85%
Fast Redact Intermediate Wild Restore | 118,97 | 96,42 | 22,55% | 101,235 | -4,81%
JSO N Stringify | 193,94 | 209,40 | -15,46% | 219,837 | -10,43%
Noir V2 Serialize | 269,23 | 317,01 | -47,78% | 259,441 | 57,57%
Fast Redact Serialize | 208,80 | 221,33 | -12,53% | 245,216 | -23,88%
Noir V2 Wild Serialize | 278,45 | 287,44 | -8,99% | 259,385 | 28,05%
Fast Redact Wild Serialize | 237,41 | 229,06 | 8,34% | 232,912 | -3,85%
Fast Redact Intermediate Wild Serialize | 303,42 | 336,36 | -32,94% | 295,602 | 40,75%
Fast Redact Intermediate Wild Match Wild Outcome Serialize | 451,95 | 477,71 | -25,76% | 448,108 | 29,60%
Fast Redact Static Match Wild Outcome Serialize | 229,21 | 219,57 | 9,64% | 227,03 | -7,46%
Noir V2 Censor Function | 69,71 | 63,62 | 6,09% | N/A | N/A
Fast Redact Censor Function | 51,17 | 52,27 | -1,10% | N/A | N/A

**Comment by davidmarkclements** [comment]
Yeah so if you're running slack, Spotify .. really anything with electron it will skew results. also fortinet, chrome.. Dropbox and many others. 

If you wanted to get more accurate disable everything, do three runs of each and take the averages 

However, no pressing need - if there was something very wrong you'd see it every time so I'm happy with this as is

File: readme.md [hunk 1]
```diff
@@ -101,17 +101,17 @@ with `undefined` values), this restriction can be bypassed by explicitly setting
...
+This is the value which overwrites redacted properties. 
 
 Setting `censor` to `undefined` will cause properties to removed as long as this is 
 the behavior of the `serializer` – which defaults to `JSON.stringify`, which does 
 remove `undefined` properties.
 
-Supplying a function as a `censor` will cause `fast-redact` to throw. This means 
-a function censor to be introduced in the future without it being a breaking change. 
+Setting `censor` to a function will cause `fast-redact` to invoke it with the original 
+value. Funciton return will be used as redacted value.
```
Comment (davidmarkclements) [inline readme.md]:
```suggestion
value. The output of the `censor` function sets the redacted value.
```

**Comment by feugy** [comment]
Thanks for the hints.
I've run the benchmark with no application started, just the terminals and the OS.
5 times each, what you see is the average.


Test | 1 – Censor Fct with cache | 2 – Censor Fct | 3 – No support | 1-2 Δ | 1-3 Δ | 2-3 Δ
-- | -- | -- | -- | -- | -- | --
Noir V2 | 88,50 | 81,31 | 89,16 | 7,18% | -0,66% | -7,84%
Fast Redact | 7,73 | 11,15 | 8,11 | -3,42% | -0,38% | 3,04%
Fast Redact Restore | 22,39 | 36,86 | 22,37 | -14,47% | 0,01% | 14,49%
Noir V2 Wild | 130,57 | 117,64 | 121,33 | 12,93% | 9,24% | -3,68%
Fast Redact Wild | 44,18 | 49,66 | 47,87 | -5,48% | -3,69% | 1,79%
Fast Redact Wild Restore | 54,95 | 76,02 | 59,35 | -21,07% | -4,40% | 16,67%
Fast Redact Intermediate Wild | 168,30 | 165,99 | 173,01 | 2,31% | -4,71% | -7,02%
Fast Redact Intermediate Wild Restore | 137,43 | 114,72 | 130,74 | 22,70% | 6,69% | -16,01%
JSON Stringify | 211,48 | 260,39 | 208,18 | -48,91% | 3,30% | 52,22%
Noir V2 Serialize | 298,91 | 354,22 | 292,55 | -55,31% | 6,36% | 61,67%
Fast Redact Serialize | 213,67 | 261,02 | 221,35 | -47,35% | -7,68% | 39,67%
Noir V2 Wild Serialize | 302,63 | 323,29 | 288,54 | -20,66% | 14,09% | 34,75%
Fast Redact Wild Serialize | 253,32 | 305,47 | 248,88 | -52,14% | 4,45% | 56,59%
Fast Redact Intermediate Wild Serialize | 333,88 | 364,81 | 329,58 | -30,94% | 4,30% | 35,24%
Fast Redact Intermediate Wild Match Wild Outcome Serialize | 474,45 | 513,68 | 453,05 | -39,23% | 21,41% | 60,63%
Fast Redact Static Match Wild Outcome Serialize | 243,37 | 278,96 | 239,48 | -35,59% | 3,88% | 39,48%
Noir V2 Censor Function | 103,78 | 84,30 | N/A | 19,49% | N/A | N/A
Fast Redact Censor Function | 77,49 | 104,74 | N/A | -27,25% | N/A | N/A
Noir V2 | 88,56 | 105,78 | 76,2636 | -17,22% | 12,30% | 29,52%
Fast Redact | 1,91 | 2,31 | 1,9212 | -0,40% | -0,01% | 0,39%
Fast Redact Restore | 12,29 | 16,14 | 11,9256 | -3,85% | 0,36% | 4,21%
Noir V2 Wild | 104,36 | 70,33 | 111,8668 | 34,04% | -7,50% | -41,54%
Fast Redact Wild | 27,50 | 25,31 | 28,2474 | 2,20% | -0,75% | -2,94%
Fast Redact Wild Restore | 45,62 | 60,44 | 57,537 | -14,82% | -11,92% | 2,90%
Fast Redact Intermediate Wild | 182,38 | 176,04 | 191,5834 | 6,35% | -9,20% | -15,55%
Fast Redact Intermediate Wild Restore | 133,67 | 101,79 | 127,4426 | 31,88% | 6,23% | -25,65%
JSON Stringify | 208,91 | 256,54 | 209,1652 | -47,63% | -0,26% | 47,37%
Noir V2 Serialize | 284,31 | 329,37 | 279,6778 | -45,06% | 4,63% | 49,69%
Fast Redact Serialize | 218,01 | 258,48 | 210,8174 | -40,47% | 7,19% | 47,66%
Noir V2 Wild Serialize | 289,35 | 292,40 | 290,4842 | -3,05% | -1,13% | 1,92%
Fast Redact Wild Serialize | 250,00 | 304,32 | 249,4932 | -54,32% | 0,50% | 54,82%
Fast Redact Intermediate Wild Serialize | 313,90 | 337,09 | 319,7814 | -23,19% | -5,88% | 17,31%
Fast Redact Intermediate Wild Match Wild Outcome Serialize | 467,46 | 486,35 | 451,8226 | -18,89% | 15,64% | 34,53%
Fast Redact Static Match Wild Outcome Serialize | 234,84 | 280,87 | 245,3556 | -46,03% | -10,52% | 35,52%
Noir V2 Censor Function | 88,06 | 70,45 | N/A | 17,61% | N/A | N/A
Fast Redact Censor Function | 73,22 | 80,52 | N/A | -7,30% | N/A | N/A



Indeed, the results are quite different:
- caching `typeof censor === 'function'` makes a huge difference (7~27% loss)
- the overhead added by the feature seems acceptable... (4~7% loss)
- ...except for intermediate wildcards (15~21% loss)

**Comment by feugy** [comment]
So gents, what do you think about this?

**Comment by feugy** [comment]
@mcollina good point!
I could move the ternary in `redactor()` and `dynamicRedactTmpl()` functions pretty easily.
I couldn't in `nestedRedact()` and `groupRedact()`, since they are not "dynamic" code (turning them to dynamic requires quite some discussion).

The performances (column `0-3 Δ` )are quite good, although I'm still not convince we could draw conclusion with such small volumes.


Test | 0 – Censor Fct with template | 1 – Censor Fct with cache | 2 – Censor Fct | 3 – No support | 0-3 Δ | 1-2 Δ | 1-3 Δ | 2-3 Δ
-- | -- | -- | -- | -- | -- | -- | -- | --
Noir V2 | 89,19 | 88,50 | 81,31 | 89,16 | 0,04% | 7,18% | -0,66% | -7,84%
Fast Redact | 8,53 | 7,73 | 11,15 | 8,11 | 0,42% | -3,42% | -0,38% | 3,04%
Fast Redact Restore | 20,53 | 22,39 | 36,86 | 22,37 | -1,84% | -14,47% | 0,01% | 14,49%
Noir V2 Wild | 106,16 | 130,57 | 117,64 | 121,33 | -15,16% | 12,93% | 9,24% | -3,68%
Fast Redact Wild | 31,06 | 44,18 | 49,66 | 47,87 | -16,81% | -5,48% | -3,69% | 1,79%
Fast Redact Wild Restore | 36,06 | 54,95 | 76,02 | 59,35 | -23,29% | -21,07% | -4,40% | 16,67%
Fast Redact Intermediate Wild | 143,35 | 168,30 | 165,99 | 173,01 | -29,66% | 2,31% | -4,71% | -7,02%
Fast Redact Intermediate Wild Restore | 112,96 | 137,43 | 114,72 | 130,74 | -17,78% | 22,70% | 6,69% | -16,01%
JSON Stringify | 197,78 | 211,48 | 260,39 | 208,18 | -10,39% | -48,91% | 3,30% | 52,22%
Noir V2 Serialize | 266,36 | 298,91 | 354,22 | 292,55 | -26,19% | -55,31% | 6,36% | 61,67%
Fast Redact Serialize | 199,44 | 213,67 | 261,02 | 221,35 | -21,91% | -47,35% | -7,68% | 39,67%
Noir V2 Wild Serialize | 258,23 | 302,63 | 323,29 | 288,54 | -30,30% | -20,66% | 14,09% | 34,75%
Fast Redact Wild Serialize | 232,21 | 253,32 | 305,47 | 248,88 | -16,66% | -52,14% | 4,45% | 56,59%
Fast Redact Intermediate Wild Serialize | 307,44 | 333,88 | 364,81 | 329,58 | -22,14% | -30,94% | 4,30% | 35,24%
Fast Redact Intermediate Wild Match Wild Outcome Serialize | 439,07 | 474,45 | 513,68 | 453,05 | -13,97% | -39,23% | 21,41% | 60,63%
Fast Redact Static Match Wild Outcome Serialize | 225,65 | 243,37 | 278,96 | 239,48 | -13,84% | -35,59% | 3,88% | 39,48%
Noir V2 Censor Function | 82,78 | 103,78 | 84,30 | N/A | N/A | 19,49% | N/A | N/A
Fast Redact Censor Function | 51,66 | 77,49 | 104,74 | N/A | N/A | -27,25% | N/A | N/A
Noir V2 | 70,03 | 88,56 | 105,78 | 76,26 | -6,23% | -17,22% | 12,30% | 29,52%
Fast Redact | 1,75 | 1,91 | 2,31 | 1,92 | -0,17% | -0,40% | -0,01% | 0,39%
Fast Redact Restore | 14,18 | 12,29 | 16,14 | 11,93 | 2,26% | -3,85% | 0,36% | 4,21%
Noir V2 Wild | 84,89 | 104,36 | 70,33 | 111,87 | -26,98% | 34,04% | -7,50% | -41,54%
Fast Redact Wild | 21,68 | 27,50 | 25,31 | 28,25 | -6,57% | 2,20% | -0,75% | -2,94%
Fast Redact Wild Restore | 32,59 | 45,62 | 60,44 | 57,54 | -24,94% | -14,82% | -11,92% | 2,90%
Fast Redact Intermediate Wild | 184,55 | 182,38 | 176,04 | 191,58 | -7,03% | 6,35% | -9,20% | -15,55%
Fast Redact Intermediate Wild Restore | 119,23 | 133,67 | 101,79 | 127,44 | -8,21% | 31,88% | 6,23% | -25,65%
JSO N Stringify | 202,69 | 208,91 | 256,54 | 209,17 | -6,48% | -47,63% | -0,26% | 47,37%
Noir V2 Serialize | 263,88 | 284,31 | 329,37 | 279,68 | -15,79% | -45,06% | 4,63% | 49,69%
Fast Redact Serialize | 198,75 | 218,01 | 258,48 | 210,82 | -12,07% | -40,47% | 7,19% | 47,66%
Noir V2 Wild Serialize | 251,84 | 289,35 | 292,40 | 290,48 | -38,64% | -3,05% | -1,13% | 1,92%
Fast Redact Wild Serialize | 229,16 | 250,00 | 304,32 | 249,49 | -20,33% | -54,32% | 0,50% | 54,82%
Fast Redact Intermediate Wild Serialize | 299,64 | 313,90 | 337,09 | 319,78 | -20,15% | -23,19% | -5,88% | 17,31%
Fast Redact Intermediate Wild Match Wild Outcome Serialize | 431,34 | 467,46 | 486,35 | 451,82 | -20,48% | -18,89% | 15,64% | 34,53%
Fast Redact Static Match Wild Outcome Serialize | 220,04 | 234,84 | 280,87 | 245,36 | -25,32% | -46,03% | -10,52% | 35,52%
Noir V2 Censor Function | 57,37 | 88,06 | 70,45 | N/A | N/A | 17,61% | N/A | N/A
Fast Redact Censor Function | 45,52 | 73,22 | 80,52 | N/A | N/A | -7,30% | N/A | N/A

**Comment by mcollina** [comment]
I can’t really read that diagram. Are we on par or still decreasing?

cc @davidmarkclements

**Comment by davidmarkclements** [comment]
I think this is fine, exaggerated variations are due to machine noise imo - also implementation-wise I don't think we would do this in any significantly different way 

thanks @feugy !

**Pull request merged**